tests/test_energy_overflows.py

```diff
@@ -0,0 +1,282 @@
+
+        # Check that all of the energies out of a batch are identical
+        _, _, energies = unbound_pot.execute_selective_batch([x0] * 2, [params] * 2, [box] * 2, False, False, True)
+        for energy in energies.reshape(-1):
+            verify_energies(selective_energy, energy)
+
+        fixed_energy, overflows = bound_pot.execute_fixed(x0, box)
+
+        # If there are no overflows, the fixed energy value will match
+        if overflows == 0:
+            assert FIXED_TO_FLOAT(fixed_energy) == selective_energy
+
+        return bound_energy
+
+    reference_energy = potential(x0, params, box)
+    test_energy = compute_potential_energy(potential)
+    np.testing.assert_allclose(reference_energy, test_energy, rtol=rtol, atol=atol)
+
+    # If we call the two potentials separately, we will now see that both return inf energies
+    all_pairs_energy = compute_potential_energy(all_pairs)
+    assert not np.isfinite(all_pairs_energy)
+
+    pair_list_energy = compute_potential_energy(pair_list)
+    assert not np.isfinite(pair_list_energy)
+
+    # Make sure that the overflows are actually cancelling out
+    pair_list_bound = pair_list.to_gpu(precision).bind(params).bound_impl
+    all_pairs_bound = all_pairs.to_gpu(precision).bind(params).bound_impl
+
+    fixed_all_pairs_energy, all_pairs_overflows = all_pairs_bound.execute_fixed(x0, box)
+    fixed_pair_list_energy, pair_list_overflows = pair_list_bound.execute_fixed(x0, box)
+    assert all_pairs_overflows != 0
+    assert pair_list_overflows != 0
+    assert all_pairs_overflows + pair_list_overflows == 0
+
+    # The sum of the fixed point energies should be bitwise identical to the test energy
+    assert FIXED_TO_FLOAT(fixed_all_pairs_energy + fixed_pair_list_energy) == test_energy
+
+
+@pytest.mark.xfail(strict=True)
+@pytest.mark.parametrize("precision", [np.float32, np.float64])
+def test_energy_overflows_with_summation_of_energies(precision):
+
+    num_atoms = 1000
+
+    # all particles evenly spaced along x-axis, in a very large box
+    spacing = 0.75
+    x = spacing * np.array([np.arange(num_atoms), np.zeros(num_atoms), np.zeros(num_atoms)]).T
+    box = np.eye(3) * 10000.0
+    assert (np.diag(box) > spacing * num_atoms).all()
+
+    nb_params = np.ones((num_atoms, 4))
+    nb_params[:, -1] = 0.0
+
+    nonbonded = Nonbonded(
+        num_atoms=num_atoms,
+        exclusion_idxs=np.array([(0, num_atoms - 1)], dtype=np.int32),
+        scale_factors=np.zeros((1, 2)),
+        beta=2.0,
+        cutoff=1.2,
+    )
+    nonbonded_gpu = nonbonded.to_gpu(precision)
+
+    # TBD: Remove the try/finally to clean up the GPU potential, else calling reset_device
+    # will fail
+    try:
+        # The reference should return a large energy value
+        assert nonbonded(x, nb_params, box) > 0.0
+
+        # The GPU potentials will overflow, resulting in a nan value
+        assert not np.isfinite(nonbonded_gpu(x, nb_params, box))
```

Comment:
[not immediately actionable]

(If needed in the future, can tweak / weaken this condition slightly: the main problem in the context of https://github.com/proteneer/timemachine/issues/481 is if `sign(nonbonded_gpu(...)) < 0` when `nonbonded_ref(...) >> 0`. Speculatively, it would also be reasonable for `nonbonded_gpu(...)` to return a number close to `nonbonded_ref(...)`, which on this input is large / maybe not representable with current fixed-point accumulator parameters, but does not _have_ to be defined as +inf...)

Reply:
For the foreseeable future, we will stick with the fixed-point accumulator, despite an earlier version of the PR description indicating that we would move to float accumulation. Which means this likely continue to be a `nan`. Does this raise potential issues for your use cases?

Reply:
Thanks -- I think that makes sense!

(Just wasn't sure if the parameters of the fixed-point accumulator might vary in the foreseeable future, e.g. to change the max or min representable energy value. If so, the energy of this input wouldn't necessarily have to be asserted to be not finite -- nan, +inf, or any large positive number would be usable. I think the main behavior that needs to be ruled out on inputs like this is to return a negative number.)